### PR TITLE
display sigma letter next to metric values

### LINKED ISSUE #364: Show the sum symbol next to the metric sum
Currently CC sums the metrics for the buildings, when hovering over folders. It would be nice to also see a sum symbol to know what it does.

### PR DESCRIPTION
**Pull request by NearW**
# sigma letter next to metric values

closes #364

## Description
- letter sigma displayed left of the metric value (not a fontawesome-icon)
- Fixed wrong interface of AttributeTypes
- AttributeTypes will no be set to `{nodes: [], edges: []}` instead of `{}`
- AttributeType are no longer 0 and 1. The enum references strings "absolute" and "relative" now

Boyscout:
- Fixed bug and added e2e-test, that ensures metric sum is shown for folders

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [x] **All** requirements mentioned in the issue are implemented
- [x] Does match the Code of Conduct and the Contribution file
- [x] Task has its own **GitHub issue** (something it is solving)
  - [x] Issue number is **included in the commit messages** for traceability
- [x] **Update the README.md** with any changes/additions made
- [x] **Update the CHANGELOG.md** with any changes/additions made
- [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [x] **All tests pass**
- [x] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [x] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [x] **Manual testing** did not fail

### CODE REVIEW

**Comment by alschmut** [comment]
@Richargh were you thinking about two different symbol?. One for an absolute sum and one for relative percentage values (e.x. metric: test-coverage). This data could potentially be stored in `fileSettings.attributeTypes` (although only the codemaatimporter provides that data)

**Comment by alexhunziker** [comment]
grafik

I prefer a smaller gap between the sum symbol and the metric, i think that way it becomes clearer to what the sum symbol is refering. What do you think?

**Comment by NearW** [comment]
@alschmut https://github.com/MaibornWolff/codecharta/issues/365
Relative metrics is another issue

**Comment by Richargh** [comment]
I have a question: is the sum symbol displayed when I hover over a single building? In the screenshot it looks like it is and I feel like that's wrong. I want the sum symbol to only be displayed when I acutally sum values. Because at first it's not there and suddenly it pops up when I hover over a directory. That should get the user thinking.

Also we can then replace the sum/sigma symbol with the average or mean symbol (if there is one) for attributes that are relative according to the attributetypes.

**Comment by alschmut** [comment]
@Richargh the current problem is, that we do not have the data from the importers to actually differnatiate between the absolute and the relative metrics. So we should (as @NearW mentioned) implement the relative-symbol together with #365

**Comment by alschmut** [comment]
@Richargh Okey I think I got it: So that means we do something like
```
if (attribiteTypes.includes(metric) && attribiteTypes[metric].isRelative)
  // show relative symbol
else 
  // show sum/sigma symbol
```
but still we would have to calculate cumulative relative values for the folders. Because we always sum it up at the moment. In my opinion the whole topic about *relative metrics* is still a separate issue 😄

**Comment by Richargh** [comment]
Relative metrics and calculating relative metrics is definately the separate issue #365. Your example-code is also part of that separate issue.

For the future: In another separate story we also need to describe all the attribute types in all importers as well or add a new `ccsh attributeFilter` that describes all attributes. 

But back to my original question: can we please only show the sum symbol, when we actually sum something? 

**When** I hover over a single building
**Then** the metric wihout a sum symbol is shown

**When** I hover over a package
**Then** the metric sum with a sum symbol is shown

**Comment by Richargh** [comment]
What's the status here?

File: visualization/app/codeCharta/codeCharta.model.ts [hunk 1]
```diff
@@ -113,17 +113,17 @@ export interface ColorRange {
 }
 
 export interface AttributeTypes {
-	nodes?: {
+	nodes?: Array<{
```
Comment (alschmut) [inline visualization/app/codeCharta/codeCharta.model.ts]: Remove the question mark?

File: visualization/app/codeCharta/state/metric.service.ts [hunk 2]
```diff
@@ -64,6 +74,20 @@ export class MetricService implements FileStateServiceSubscriber, SettingsServic
 		return metric ? metric.maxValue : undefined
 	}
 
+	public getAttributeTypeByMetric(metricName: string, settings: Settings): AttributeType {
+		let attributeType: AttributeType = null
```
Comment (alschmut) [inline visualization/app/codeCharta/state/metric.service.ts]: Can we refactor these loops to something like the following?
```
return settings.fileSettings.attributeTypes.nodes
    .find(x => Object.keys(x)[0] === metricName)
```

File: visualization/app/codeCharta/state/settings.service.ts [hunk 3]
```diff
@@ -13,41 +13,39 @@ export interface SettingsServiceSubscriber {
...
+	private readonly debounceBroadcast: () => void
 
 	constructor(private $rootScope: IRootScopeService, private $timeout: ITimeoutService, private loadingGifService: LoadingGifService) {
 		this.settings = this.getDefaultSettings()
-		this.debounceBroadcast = _.debounce(
-			(update: RecursivePartial<Settings>) =>
-				this.$rootScope.$broadcast(SettingsService.SETTINGS_CHANGED_EVENT, { settings: this.settings, update: update }),
-			400
-		)
+		this.debounceBroadcast = _.debounce(() => {
```
Comment (alschmut) [inline visualization/app/codeCharta/state/settings.service.ts]: Do we wanna separate the two broadcast lines into an extra function?

File: visualization/app/codeCharta/state/settings.service.ts [hunk 4]
```diff
@@ -13,41 +13,39 @@ export interface SettingsServiceSubscriber {
...
-		})
-		this.updateSettings({
-			fileSettings: this.getNewFileSettings(fileStates)
-		})
+		this.resetDynamicAndFileSettings(fileStates)
 	}
 
-	public onImportedFilesChanged(fileStates: FileState[], event: angular.IAngularEvent) {}
+	public onImportedFilesChanged(fileStates: FileState[], event: angular.IAngularEvent) {
+		this.resetDynamicAndFileSettings(fileStates)
```
Comment (alschmut) [inline visualization/app/codeCharta/state/settings.service.ts]: Is this necessary, as the `onFileSelectionStatesChanged`-event always gets called after new files were imported?

File: visualization/app/codeCharta/ui/metricType/metricType.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+			expect(CodeMapMouseEventService.subscribe).toHaveBeenCalledWith($rootScope, metricTypeController)
+		})
+	})
+
+	describe("onSettingsChanged", () => {
+		it("should set the areaMetricType to absolute", () => {
+			metricTypeController.onSettingsChanged(settings, { dynamicSettings: { areaMetric: "rloc" } }, undefined)
+
+			expect(metricTypeController["_viewModel"].areaMetricType).toBe("absolute")
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/metricType/metricType.component.spec.ts]: Use of interface `AttributeType.absolute`?

File: visualization/app/codeCharta/util/settingsMerger.spec.ts [hunk 6]
```diff
@@ -126,106 +126,40 @@ describe("SettingsMerger", () => {
 		})
 	})
 
-	describe("MarkedPackages merge", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/util/settingsMerger.spec.ts]: Why were all these MarkedPackages tests deleted?

File: visualization/app/codeCharta/state/metric.service.ts (hunk 2 above)
Comment (NearW) [reply visualization/app/codeCharta/state/metric.service.ts]: refactored to
```
public getAttributeTypeByMetric(metricName: string, settings: Settings): AttributeType {
		const attributeType = settings.fileSettings.attributeTypes.nodes.find(x => _.findKey(x) === metricName)

		if (attributeType) {
			return attributeType[metricName]
		}
		return null
	}
```

File: visualization/app/codeCharta/state/settings.service.ts (hunk 4 above)
Comment (NearW) [reply visualization/app/codeCharta/state/settings.service.ts]: If that is the case, no. removed.

File: visualization/app/codeCharta/ui/metricType/metricType.component.spec.ts (hunk 5 above)
Comment (NearW) [reply visualization/app/codeCharta/ui/metricType/metricType.component.spec.ts]: done

File: visualization/app/codeCharta/util/settingsMerger.spec.ts (hunk 6 above)
Comment (NearW) [reply visualization/app/codeCharta/util/settingsMerger.spec.ts]: Good catch. I can't explain why this happened. I re-added them

File: visualization/app/codeCharta/state/settings.service.ts (hunk 3 above)
Comment (NearW) [reply visualization/app/codeCharta/state/settings.service.ts]: Moved the broadcast and reset of `this.update` into a function `notifySubscribers`. I couldn't find a better fitting name and I think `notifySubscribersAndResetUpdateState` is too much

**Pull request merged**